src/Build/Instance/ProjectInstance.cs

```diff
@@ -487,6 +487,38 @@ private ProjectInstance(Project linkedProject, bool fastItemLookupNeeded)
             _explicitToolsVersionSpecified = linkedProject.SubToolsetVersion != null;
 
             _isImmutable = true;
+
+            static List<string> GetImportFullPaths(ObjectModelRemoting.ProjectLink projectLink)
+            {
+                // Imports collection contains ResolvedImports, which are structures which are much bigger than the string list.
+                // We convert to the string list and let Imports itself to BE GCed.
```

Comment:
In CPS we can implement CachedProject.Imports as a collection that implements both IList<ResolvedImport> and IList<string>.  This would let us avoid instantiating any ResolvedImport objects in the first place and instead only allocate List<string> that we actually want. (But for compat and to satisfy the interface we would still have the ability to act as an IList<ResolvedImport> that lazily constructs ResolvedImport objects if requested).